### PR TITLE
TinyCrypt deprecation - library's usage removal part 1

### PR DESCRIPTION
**Pull request by valeriosetti**
This is the first step toward the removal of TinyCrypt library usage from the Zephyr's codebase. This is the continuation of the initial PR (https://github.com/zephyrproject-rtos/zephyr/pull/79566) that started the deprecation of the TinyCrypt library.

The idea is to do basic/simple removals here and keep BT stuff for a follow-up separated PR.

This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/79566. Initial commit is taken from there.

### CODE REVIEW

**Review by d3zd3z — DISMISSED** [review]
This looks good. I'll be curious if we get any balking about it.

**Review by dkalowsk — DISMISSED** [review]
Thanks for the work here.  Looks good to me!

**Comment by valeriosetti** [comment]
@dkalowsk @d3zd3z sorry I had to rebase to solve a conflict in a test yaml file and this unfortunately dismissed your review. Can you please take another look at this PR?

**Review by tomi-font — CHANGES_REQUESTED** [review]
Let's just make sure that any of this can actually be done before 4.2, as the TinyCrypt library was just now marked as deprecated for 4.0.
@keith-zephyr?

**Comment by ceolin** [comment]
> Let's just make sure that any of this can actually be done before 4.2, as the TinyCrypt library was just now marked as deprecated for 4.0. @keith-zephyr?

Yep ! @valeriosetti we have to follow the deprecation process (https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated). So we have to keep it around a little bit more. We can though remove internal usages like the one you did for the random subsystem. 

I think the tinycrypt tests will need to stay as well while the library is deprecated.

**Review by ceolin — CHANGES_REQUESTED** [review]
You have to mark the shim driver as deprecated for now and remove it together with the library. The changes in `jwt` and `random` can get in.

**Comment by dkalowsk** [comment]
> Let's just make sure that any of this can actually be done before 4.2, as the TinyCrypt library was just now marked as deprecated for 4.0. @keith-zephyr?

@tomi-font likely need to make an issue to remove Tinycrypt and tie it to the 4.2 milestone so we don't forget that it's there.

**Comment by tomi-font** [comment]
> @tomi-font likely need to make an issue to remove Tinycrypt and tie it to the 4.2 milestone so we don't forget that it's there.

Yeah, actually this is already done as part of the umbrella issue for 4.2: https://github.com/zephyrproject-rtos/zephyr/issues/77070

**Comment by valeriosetti** [comment]
> You have to mark the shim driver as deprecated for now and remove it together with the library. The changes in `jwt` and `random` can get in.

Ops, sorry, I thought that keeping the library itself available as module, but removing _all_ usages was fine. Btw I reverted some commits from the history, so the PR should be ready for reviews again (assuming the CI approves, of course :) )

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [inline doc/releases/migration-guide-4.0.rst:200]:
```suggestion
* Following the deprecation of the TinyCrypt library (:github:`79566`), the TinyCrypt-based
  shim driver was marked as deprecated. (:github:`79653`)
```

**Comment by tomi-font** [comment]
> You have to mark the shim driver as deprecated for now and remove it together with the library. The changes in jwt and random can get in.

@ceolin Are you sure about the latter part? I can imagine users currently using TinyCrypt for those may face breaking changes when they are removed (which means as soon as 4.0 if we remove them right away) due to e.g. using TinyCrypt and not Mbed TLS (and being already tight on size constraints). I don't know. Just trying to make sure this is done right.

File: drivers/crypto/Kconfig
Comment (valeriosetti) [reply drivers/crypto/Kconfig:25]: Yeah, I looked in the codebase and there seems to be no standard pattern for this: in some cases the label reports "[DEPRECATED]", but not all. Anyway, it does not hurt and it makes things clearer, so I will add it. Thanks.

PS = Too bad that `menuconfig` does not automatically add that label when `select DEPRECATED` is set for a certain Kconfig.

**Comment by valeriosetti** [comment]
> > You have to mark the shim driver as deprecated for now and remove it together with the library. The changes in jwt and random can get in.
> 
> @ceolin Are you sure about the latter part? I can imagine users currently using TinyCrypt for those may face breaking changes when they are removed (which means as soon as 4.0 if we remove them right away) due to e.g. using TinyCrypt and not Mbed TLS (and being already tight on size constraints). I don't know. Just trying to make sure this is done right.

Giving my 2 cents here while waiting for Flavio's reply. I think the answer is "yes" and the reason is (https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated):
 
> - Code using the deprecated API needs to be modified to remove usage of said API

**Review by d3zd3z — DISMISSED** [review]

**Comment by dkalowsk** [comment]
@ceolin and @tomi-font isn't this needed for the 4.0 release?  AKA we're in RC1 right now, what needs to be done to make this into RC2?

**Comment by ceolin** [comment]
> > > You have to mark the shim driver as deprecated for now and remove it together with the library. The changes in jwt and random can get in.
> > 
> > 
> > @ceolin Are you sure about the latter part? I can imagine users currently using TinyCrypt for those may face breaking changes when they are removed (which means as soon as 4.0 if we remove them right away) due to e.g. using TinyCrypt and not Mbed TLS (and being already tight on size constraints). I don't know. Just trying to make sure this is done right.
> 
> Giving my 2 cents here while waiting for Flavio's reply. I think the answer is "yes" and the reason is (https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated):
> 
> > * Code using the deprecated API needs to be modified to remove usage of said API

Yep. We have to keep it for external use cases but internally we can remove it.

**Review by ceolin — DISMISSED** [review]

**Comment by ceolin** [comment]
> @ceolin and @tomi-font isn't this needed for the 4.0 release? AKA we're in RC1 right now, what needs to be done to make this into RC2?

These changes look good to me :)

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [inline doc/releases/migration-guide-4.0.rst:554]:
```suggestion
* By default, the signature is now computed using the PSA Crypto API for both RSA and ECDSA.
```

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [inline doc/releases/migration-guide-4.0.rst:557]:
```suggestion
  following the deprecation of the TinyCrypt library (:github:`79566`), usage
```

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [inline doc/releases/migration-guide-4.0.rst:563]:
```suggestion
  * :kconfig:option:`JWT_SIGN_RSA_PSA` (default) RSA signature using the PSA Crypto API;
```
and same for `JWT_SIGN_ECDSA_PSA`

File: subsys/jwt/Kconfig
Comment (tomi-font) [inline subsys/jwt/Kconfig:26]:
```suggestion
	bool "Use RSA signature (RS-256). Use PSA Crypto API."
```

File: subsys/jwt/Kconfig
Comment (tomi-font) [inline subsys/jwt/Kconfig:35]:
```suggestion
	bool "Use ECDSA signature (ES-256). Use PSA Crypto API."
```

File: subsys/jwt/jwt_psa.c
Comment (tomi-font) [inline subsys/jwt/jwt_psa.c:35]:
```suggestion
#endif /* CONFIG_JWT_SIGN_ECDSA_PSA */
```

**Review by tomi-font — CHANGES_REQUESTED** [review]
I only have one blocking comment: https://github.com/zephyrproject-rtos/zephyr/pull/79653#discussion_r1822499582

File: doc/releases/migration-guide-4.0.rst
Comment (valeriosetti) [reply doc/releases/migration-guide-4.0.rst:565]: That's correct. I didn't intentionally add the previous symbols (those introduced in #78243) because they were never part of any release if this PR had make it to be merged before 4.0 code freeze. However this is not the case anymore, so yes, I need to update the documentation as well :)

File: doc/releases/migration-guide-4.0.rst
Comment (valeriosetti) [reply doc/releases/migration-guide-4.0.rst:555]: Mmmm albeit what you suggest is the "standard" way of adding github references, that works fine if the reference is at the end of the sentence, which is not the case here. I prefer to keep that full stop mark and remove the previous one because that reference is related to the previous sentence, not the one that is starting.

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [reply doc/releases/migration-guide-4.0.rst:565]: Maybe there's still a small chance for this to go in for 4.0. Dunno. But also, the original JWT Kconfig options that were removed, those that you did not introduce yourself, are they documented in the migration guide?

File: subsys/jwt/jwt_psa.c
Comment (valeriosetti) [reply subsys/jwt/jwt_psa.c:35]: Well this was related to the fact that I wrote something like:
```
#if defined(CONFIG_JWT_SIGN_ECDSA_PSA)
...
#else  /* CONFIG_JWT_SIGN_RSA_PSA */
...
#endif /* CONFIG_JWT_SIGN_ECDSA_PSA || CONFIG_JWT_SIGN_RSA_PSA */
```
The idea was to state the you surely picked one of the two. But since this created confusion I will fix this

File: doc/releases/migration-guide-4.0.rst
Comment (tomi-font) [reply doc/releases/migration-guide-4.0.rst:555]: My point here is that you have two dots, without sentence between them.

File: subsys/jwt/jwt_psa.c
Comment (tomi-font) [reply subsys/jwt/jwt_psa.c:35]: Yeah I understood that, but it's rather done this way:
```
#if FIRST_COND
[...]
#else /* SOME_OTHER_COND */
[...]
#endif /* FIRST_COND */
```
Not saying it's the best, but just for consistency across the codebase.

File: doc/releases/migration-guide-4.0.rst
Comment (valeriosetti) [reply doc/releases/migration-guide-4.0.rst:565]: > Maybe there's still a small chance for this to go in for 4.0

I don't know either. If not, then I will address this chance when the new migration-guide document will be added to the codebase.

>  But also, the original JWT Kconfig options that were removed, those that you did not introduce yourself, are they documented in the migration guide?

#78243 documented new symbols [added there](https://github.com/zephyrproject-rtos/zephyr/pull/78243/files#diff-ada0effa18bf591a10f0e71db409689e026c8c35cd790a1b0026d25cac005535R364), but no, this one lacks documentation about the removal

**Review by tomi-font — APPROVED** [review]
Looks good to me! I realized we have one more reason to want this in for 4.0. (@dkalowsk @mmahadevan108)
This is replacing Kconfig options in JWT that themselves replaced older Kconfig options, within this release.
What this means if we only merge this for 4.1, is that there will be a few Kconfig options that are introduced in 4.0 and removed in 4.1. Really not great. Merging this for 4.0 makes this double change atomic, so I hope we can do that.
Also, this helps with the deprecation of TinyCrypt as this deprecates the shim driver (among others), which makes sense to have in 4.0 as well.
@ceolin @d3zd3z please re-review ASAP

**Comment by mmahadevan108** [comment]
I mentioned this PR in the Oct 30th TSC meeting for an exception so we can merge and include in the 4.0 release. I have also posted this PR for discussion in the release channel.

**Review by dkalowsk — APPROVED** [review]
Looks right to me.  @ceolin looking for your review as well.

One minor question, should the MAINTAINERS.yml file status be updated to also note that TinyCrypt is deprecated and not getting "odd fixes"?

**Review by ceolin — APPROVED** [review]
re-approving it

**Review by dkalowsk — COMMENTED** [review]
I don't see anything in the PR that should block it (and it has my approval just not giving it in GitHub yet due the next steps).  My understanding from conversations in Discord has been that this part 1 really needs to be merged with part 2.  Even then there are some concerns that BT isn't quite ready for Tinycrypt to go away.  @jhedberg can you confirm my understanding here?

**Comment by jhedberg** [comment]
> My understanding from conversations in Discord has been that this part 1 really needs to be merged with part 2. Even then there are some concerns that BT isn't quite ready for Tinycrypt to go away. @jhedberg can you confirm my understanding here?

I'd say it's actually three PRs that go together, starting with #79566 which was already merged. Ultimately this is something for the release engineering team and possibly the TSC to decide, i.e. whether to make an exception or not to the practice of not having upstream users of deprecated APIs when a release is made. If an exception isn't made, then my interpretation is that #79566 should be reverted.

**Comment by mmahadevan108** [comment]
After discussion in the release meeting on Nov 5th, it was decided to merge this PR for the 4.0 release.

**Pull request merged**
